File: python/fan-out/fan-out/handler_test.py
from types import SimpleNamespace

try:
    from . import handler as h
except ImportError:
    import handler as h


class FakeResponse:
    status_code = 202
    text = ""
    headers = {"X-Call-Id": "call-1"}


def event(body, callback=""):
    headers = {"X-Callback-Url": callback} if callback else {}
    return SimpleNamespace(body=body.encode(), headers=headers)


def test_submits_each_url_with_callback(monkeypatch):
    submitted = []

    def fake_post(url, data, headers, timeout):
        submitted.append((url, data.decode(), headers, timeout))
        return FakeResponse()

    monkeypatch.setattr(h.requests, "post", fake_post)
    response = h.handle(
        event(
            "https://one.example\nhttps://two.example\n",
            "http://gateway.openfaas:8080/function/printer",
        ),
        {},
    )

    assert response["statusCode"] == 200
    assert response["body"]["submitted"] == 2
    assert response["body"]["callback"] is True
    assert len(submitted) == 2
    assert submitted[0][2]["X-Callback-Url"].endswith("/function/printer")


def test_rejects_empty_batch():
    response = h.handle(event("\n\n"), {})
    assert response["statusCode"] == 400


def test_reports_submission_failure(monkeypatch):
    failed = FakeResponse()
    failed.status_code = 503
    failed.text = "queue unavailable"
    monkeypatch.setattr(h.requests, "post", lambda *args, **kwargs: failed)

    response = h.handle(event("https://one.example\n"), {})
    assert response["statusCode"] == 502

File: node/fan-out/fan-out/handler.test.js
'use strict'

const assert = require('node:assert/strict')
const { afterEach, test } = require('node:test')

const handler = require('./handler')

const originalFetch = global.fetch

afterEach(() => {
  global.fetch = originalFetch
})

test('submits one asynchronous invocation per URL', async () => {
  const requests = []
  global.fetch = async (url, options) => {
    requests.push({ url, options })
    return new Response('', {
      status: 202,
      headers: { 'X-Call-Id': `call-${requests.length}` }
    })
  }

  const context = new TestContext()
  await handler({
    body: 'https://one.example\nhttps://two.example\n',
    headers: {}
  }, context)

  assert.equal(context.statusCode, 200)
  assert.equal(context.body.submitted, 2)
  assert.deepEqual(context.body.call_ids, ['call-1', 'call-2'])
  assert.equal(requests[0].options.body, 'https://one.example')
})

test('forwards the callback URL to every invocation', async () => {
  const requests = []
  global.fetch = async (url, options) => {
    requests.push(options)
    return new Response('', { status: 202 })
  }

  const context = new TestContext()
  await handler({
    body: 'https://one.example',
    headers: {
      'x-callback-url': 'https://results.example/callback'
    }
  }, context)

  assert.equal(context.body.callback, true)
  assert.equal(
    requests[0].headers['X-Callback-Url'],
    'https://results.example/callback'
  )
})

test('returns bad gateway when a submission is rejected', async () => {
  global.fetch = async () => new Response('queue unavailable', {
    status: 503
  })

  const context = new TestContext()
  await handler({ body: 'https://one.example', headers: {} }, context)

  assert.equal(context.statusCode, 502)
  assert.match(context.body, /queue unavailable/)
})

class TestContext {
  status (value) {
    this.statusCode = value
    return this
  }

  headers (value) {
    this.headerValues = value
    return this
  }

  succeed (value) {
    this.body = value
    return value
  }
}
